src/classes/CW_AudienceGroupControllerTest.cls

```diff
@@ -0,0 +1,333 @@
+        insert listGroups;
+        
+        User thisUser = [SELECT Id FROM User WHERE Id = :UserInfo.getUserId()];
+    	
+    	Test.startTest();
+    	
+		System.runAs(thisUser)
+	    {  
+	    	audienceGroupPageSizeRecord.Integer_Value__c = 5;
+			update audienceGroupPageSizeRecord;
+			
+	    	Test.setCurrentPage(Page.CW_Campaign_Information);
+			CW_WizardController wizCtrl = new CW_WizardController(new ApexPages.StandardController(camp));
+			wizCtrl.next();
+			
+			Test.setCurrentPage(Page.CW_Audience);
+			Apexpages.currentPage().getParameters().put('id',camp.Id);
+			CW_AudienceGroupController agCtrl = new CW_AudienceGroupController();
+			
+			agCtrl.selectAll = true;
+			agCtrl.selectDeselectAllItems();
+			system.assertEquals(true,agCtrl.getHasNext());
+			
+			agCtrl.doNext();
+			
+			system.assertEquals(agCtrl.selectAll, false);
+			
+			system.assertEquals(true,agCtrl.getHasPrevious());
+			agCtrl.doPrevious();
+			
+			agCtrl.contextItem = agCtrl.campGroups[0].aGroup.Id;
+			agCtrl.doSelectItem();
+			agCtrl.doDeselectItem();
+			
+			agCtrl.last();
+			system.AssertEquals(agCtrl.setGroups.getRecords().size(),5);
+			
+			agCtrl.first();
+			system.AssertEquals(agCtrl.setGroups.getRecords().size(),5);
+			
+			agCtrl.doNext();
+			system.AssertEquals(agCtrl.hasFirstPage,true);
+			system.AssertEquals(agCtrl.hasLastPage,true);
+			
+			agCtrl.doPrevious();
+			system.AssertEquals(agCtrl.hasFirstPage,false);
+			system.AssertEquals(agCtrl.hasLastPage,true);
+			
+	    }
+	}
+	
+	static testMethod void editSaveGroupsTest() 
+    {
+        initCustomSettings();
+        
+        audienceGroupPageSizeRecord.Integer_Value__c = null;
+		update audienceGroupPageSizeRecord;
+			
+        CW_Campaign__c camp = TestDataUtil.createCampaign();
+        insert camp;
+        
+        List<CW_Audience_Group__c> listGroups = TestDataUtil.createAudienceGroups(camp.Id,5);
+        insert listGroups;
+        
+        User thisUser = [SELECT Id FROM User WHERE Id = :UserInfo.getUserId()];
+    	
+    	Test.startTest();
+    	
+		System.runAs(thisUser)
+	    {  
+	        Test.setCurrentPage(Page.CW_Campaign_Information);
+			CW_WizardController wizCtrl = new CW_WizardController(new ApexPages.StandardController(camp));
+			wizCtrl.next();
+			
+			Test.setCurrentPage(Page.CW_Audience);
+			Apexpages.currentPage().getParameters().put('id',camp.Id);
+			CW_AudienceGroupController agCtrl = new CW_AudienceGroupController();
+			
+			system.AssertEquals(agCtrl.setGroups.getPageSize(),100);
+        	agCtrl.getFields();
+			agCtrl.editGroups();
+			
+			system.assertEquals(agCtrl.inputMode,true);
+			
+			agCtrl.cancel();
+			
+			system.assertEquals(agCtrl.inputMode,false);
+			
+			agCtrl.editGroups();
+			
+			agCtrl.getAudienceGroup();
+			
+			agCtrl.campGroups[0].aGroup.Description__c = 'Update Description';
+			agCtrl.campGroups[1].aGroup.Name = 'Update Name';
+			agCtrl.saveGroups();
+			
+			listGroups = [SELECT Id,Name,Description__c
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.